metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -906,12 +906,35 @@ private void appendRaftEvent(String name, Runnable runnable) {
                 if (this != metaLogListener) {
                     log.debug("Ignoring {} raft event from an old registration", name);
                 } else {
-                    runnable.run();
+                    try {
+                        runnable.run();
+                    } finally {
+                        maybeCompleteAuthorizerInitialLoad();
+                    }
                 }
             });
         }
     }
 
+    private void maybeCompleteAuthorizerInitialLoad() {
+        if (!needToCompleteAuthorizerLoad) return;
+        OptionalLong highWatermark = raftClient.highWatermark();
+        if (highWatermark.isPresent()) {
+            if (lastCommittedOffset + 1 >= highWatermark.getAsLong()) {
```

Comment:
I guess the only issue with this is that the high watermark is a moving target. It probably works ok since writes to the metadata log should be infrequent. Not sure I have any better ideas. Maybe we could refresh the high watermark value only once every second or something like that.

Reply:
yeah, I think this is OK for now. if we have issues we can revisit...